Compute variance and standard error

jjgoings opened the pull request:
Adds functionality to compute the variance and standard error of an expectation value.

- Adds `get_variance` function to `Simulator` to compute the variance of expectation value
- Adds `get_standard_error` function to `Simulator` to compute the standard error from the mean (shot-dependent)
- Adds simple analytical test to check variance is computed correctly

jjgoings commented on tangelo/linq/simulator_base.py:
```diff
@@ -284,6 +398,54 @@ def _get_expectation_value_from_statevector(self, qubit_operator, state_prep_cir
+                variance += 0.
+                continue
+
+            if not self.n_shots:
```
I don't think this is true. It's fine for use in computing standard error, but variance can still be non-zero with statevector. I'm not sure the best way to compute variance here in the exact case. Do Var(H) = <H^2> - <H>^2 directly?

AlexandreF-1qbit replied:
Can something like https://github.com/goodchemistryco/Tangelo/blob/7a32ef2e72f24fd393b010e0a8500da4afae6188/tangelo/linq/simulator_base.py#L347
could be used in that situation? I mean just converting the statevector to ideal frequencies, and then resuse `_get_variance_from_frequencies` to get the ideal variance.

jjgoings replied:
Looks like we didn't need extra logic! Everything is computed the same way with frequencies now, either ideal or re-sampled depending on n_shots.

AlexandreF-1qbit reviewed (commented):
I just left one comment in the thread. Beside this, I am all good with the code (I will let you make changes if necessary and mark it as approved soon after that)!

ValentinS4t1qbit commented on tangelo/linq/simulator_base.py:
```diff
@@ -231,6 +265,86 @@ def get_expectation_value(self, qubit_operator, state_prep_circuit, initial_stat
+            qubit_operator(openfermion-style QubitOperator class): a qubit
+                operator.
+            state_prep_circuit: an abstract circuit used for state preparation.
+
```
I think we need to document `initial_statevector` as well

ValentinS4t1qbit commented on tangelo/linq/simulator_base.py:
```diff
@@ -231,6 +265,86 @@ def get_expectation_value(self, qubit_operator, state_prep_circuit, initial_stat
+        """
+        variance = self.get_variance(qubit_operator, state_prep_circuit, initial_statevector)
+
+        if not self.n_shots:
```
Python offers its own version of the "ternary if", which is pretty concise and natural to read:
`return np.sqrt(variance/self.n_shots) if self.n_shots else 0.`

ValentinS4t1qbit commented on tangelo/linq/simulator_base.py:
```diff
@@ -231,6 +265,86 @@ def get_expectation_value(self, qubit_operator, state_prep_circuit, initial_stat
+            qubit_operator(openfermion-style QubitOperator class): a qubit
+                operator.
+            state_prep_circuit: an abstract circuit used for state preparation.
+
```
`initial_statevector` should be documented

ValentinS4t1qbit commented on tangelo/linq/simulator_base.py:
```diff
@@ -284,6 +398,54 @@ def _get_expectation_value_from_statevector(self, qubit_operator, state_prep_cir
+                operator.
+            state_prep_circuit: an abstract circuit used for state preparation
+                (only pure states).
+
```
initial_statevector doc

ValentinS4t1qbit commented on tangelo/linq/simulator_base.py:
```diff
@@ -284,6 +398,54 @@ def _get_expectation_value_from_statevector(self, qubit_operator, state_prep_cir
+            if len(term) > n_qubits:  # Cannot have a qubit index beyond circuit size
+                raise ValueError(f"Size of operator {qubit_operator} beyond circuit width ({n_qubits} qubits)")
+            elif not term:  # Empty term: no simulation needed
+                variance += 0.
```
I don't understand why this instruction is here. This does nothing, so it does not need to be in the code at all, we can just `continue`.

btw Python has the `pass` keyword for when we do need an instruction that does nothing, for example a class / function that needs at least an instruction so that Python does not complain. If you however have a conditional branch that just says `pass` then it's very likely you don't need that branch at all in the first place.

ValentinS4t1qbit commented on tangelo/linq/simulator_base.py:
```diff
@@ -284,6 +398,54 @@ def _get_expectation_value_from_statevector(self, qubit_operator, state_prep_cir
+
+            if not self.n_shots:
+                # Without shots, variance using statevector is identically zero
+                variance += 0.
```
If there are no shots we do nothing. So why don't we just test for the opposite case (`if self.n_shots`) and only do what's needed there? Then we can completely delete L431-433 I believe. 

It seems however further conversation may be needed about what this branch should actually do.

ValentinS4t1qbit commented on tangelo/linq/simulator_base.py:
```diff
@@ -326,6 +488,50 @@ def _get_expectation_value_from_frequencies(self, qubit_operator, state_prep_cir
+            qubit_operator(openfermion-style QubitOperator class): a qubit
+                operator.
+            state_prep_circuit: an abstract circuit used for state preparation.
+
```
initial_statevector

ValentinS4t1qbit commented on tangelo/linq/simulator_base.py:
```diff
@@ -326,6 +488,50 @@ def _get_expectation_value_from_frequencies(self, qubit_operator, state_prep_cir
+            if len(term) > n_qubits:
+                raise ValueError(f"Size of operator {qubit_operator} beyond circuit width ({n_qubits} qubits)")
+            elif not term:  # Empty term: no simulation needed
+                variance += 0.
```
remove this line, it is not necessary

ValentinS4t1qbit commented on tangelo/linq/tests/test_simulator.py:
```diff
@@ -147,6 +147,43 @@ def test_get_exp_value_mixed_state(self):
             results[b] = sim.get_expectation_value(op1, circuit_mixed)
             np.testing.assert_almost_equal(results[b], reference, decimal=2)
 
+    def test_get_variance(self):
```
I'm not sure I understand the point of testing `get_expectation_value` here.
You are not reusing those values anyway, as we recompute a brand new value in `get_variance` (exact, or from shots).

Maybe this was a little sanity check on your side to help with debugging and development ?
All I currently see is tests that do not test anything that was not tested already, unrelated to your feature, with each a probably to break depending on the samples.

ValentinS4t1qbit replied:
I think we should however test whatever that code is not using (e.g `_get_variance_from_frequencies` or `_get_variance_from_statevector`) explicitly to make sure both the underlying functions you have written are behaving as you'd expect.

jjgoings replied:
Since statevector is still based on ideal frequencies, we can handle all the functionality with `from_frequencies`.

jjgoings replied:
And those tests should be gone now :)

ValentinS4t1qbit commented on tangelo/linq/tests/test_simulator.py:
```diff
@@ -147,6 +147,43 @@ def test_get_exp_value_mixed_state(self):
+        # prepares sqrt(2/3)|0> + -i*sqrt(1/3)|1>
+        circuit = Circuit([Gate("RX", 0, parameter=2*np.arcsin(np.sqrt(1/3)))])
+
+        for b in installed_simulator:
```
@AlexandreF-1qbit @JamesB-1qbit Since all the variance functions  have been implemented in `SimulatorBase` and rely on the existing and thoroughly tested `get_expectation_value` method each backend implements: is there any meaning in testing this for all simulators ?

Couldn't we just use the default cirq simulator for that test ? We can then maybe crank it up to 10**6 shots to guarantee the comparison up to 2 decimals never fails because of "bad luck" (happens on rare occasions in other tests for me in my recent developments).

jjgoings replied:
We use just the cirq simulator now.

jjgoings commented:
@AlexandreF-1qbit should be good now! Tests for both n_shots = None and n_shots = 10**5 should pass.

ValentinS4t1qbit reviewed (changes requested):
Thank you for your PR Joshua, really appreciate it ! You did a great job with adopting our coding guidelines and seemed to have found the right place to add your feature.

 I found some minor things, that should be fairly easy to resolve / clean up.

There's code that may be redundant (reused from expectation value functions) but I'm not sure yet whether we should refactor this. For now I think it's better to address what has been brought up and resolve ongoing conversations: then it may become more obvious.

ValentinS4t1qbit commented:
> @AlexandreF-1qbit should be good now! Tests for both n_shots = None and n_shots = 10**5 should pass.

Dammit, I just gave you a review but it may mean it was given on code that's now outdated. Hopefully the stuff is still relevant

jjgoings commented:
@ValentinS4t1qbit I believe I addressed all your points -- thanks! Let me know if there is anything more I can do.
I wanted to bring up two points for your consideration:

1. Regarding refactoring, I totally agree this code is very similar to the `get_expectation`, as that's what I used as a template. I wonder if in the future we could have a function like `get_moment` where you can pass which [moment](https://en.wikipedia.org/wiki/Moment_(mathematics)) you want to compute for mean, variance, skewness, etc.
2. The functions here for variance assumes that covariance between Pauli products is zero. This is true if each term of a Hamiltonian is measured with an independent circuit. But when you group terms to reduce the number of circuits, this is not necessarily true. Depending on how you group the terms, this can make the variance increase (if Cov(Pa,Pb) > 0) or decrease (Cov(Pa,Pb) < 0). I'm not sure how to handle this, because it seems like a case-by-case basis? One fascinating pathological example was given in Sec B.2 of [this paper](https://arxiv.org/abs/1509.04279) (starts on page 15), also related in Sec IX.A in [this paper](https://par.nsf.gov/servlets/purl/10286356). Turns out that although you can partition the expectation into two measurements, it will still require more shots (33% more) to achieve the same precision as the seemingly sub-optimal three measurement partitioning. Anyway, don't really have a suggestion here other than to be aware and maybe in the future we can support computing variances given a QWC group in post-processing.

ValentinS4t1qbit reviewed (approved):
Thank you for contributing this feature to Tangelo Josh, we really appreciate it.
Congratulations on your 1st PR to this project !
